### PR TITLE
[TS-189] wrapper for UUID functions

### PR DESCRIPTION
**Pull request by jenchik**

### CODE REVIEW

File: helpers/uuid.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"fmt"
+
+	"github.com/satori/go.uuid"
+)
+
+// GenerateUUIDv4AsString returns random generated canonical string representation of UUID:
+// xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.
+func GenerateUUIDv4AsString() string {
```
Comment (im-kulikov) [inline helpers/uuid.go]: why not NewUUID?

where `GenerateUUIDv1AsString`, `GenerateUUIDv2AsString`, `GenerateUUIDv3AsString`...?

File: helpers/uuid.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	return uuid.NewV4().String()
+}
+
+// VerifyUUIDv4String verification UUID as string for conformity with version 4
+func VerifyUUIDv4String(src string) bool {
```
Comment (im-kulikov) [inline helpers/uuid.go]: for what you're using versioning? we use only v4

File: helpers/uuid.go (hunk 1 above)
Comment (jenchik) [reply helpers/uuid.go]: in more detail

File: helpers/uuid_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
+package helpers_test
```
Comment (im-kulikov) [inline helpers/uuid_test.go:1]: why you test wrappers?

File: helpers/uuid.go (hunk 2 above)
Comment (jenchik) [reply helpers/uuid.go]: yes

File: helpers/uuid_test.go (hunk 3 above)
Comment (jenchik) [reply helpers/uuid_test.go:1]: as blackbox, we do not care what we use inside

Comment (im-kulikov) [reply helpers/uuid_test.go:1]: for what?

File: helpers/uuid.go (hunk 2 above)
Comment (im-kulikov) [reply helpers/uuid.go]: "yes" what?

File: helpers/uuid.go (hunk 1 above)
Comment (im-kulikov) [reply helpers/uuid.go]: explain?

File: helpers/uuid_test.go (hunk 3 above)
Comment (jenchik) [reply helpers/uuid_test.go:1]: it is easy to change the internal implementation

File: helpers/uuid.go (hunk 2 above)
Comment (jenchik) [reply helpers/uuid.go]: always see that we are working with version 4

Comment (im-kulikov) [reply helpers/uuid.go]: if we use v4 only, why not VerifyUUID?

File: helpers/uuid_test.go (hunk 3 above)
Comment (im-kulikov) [reply helpers/uuid_test.go:1]: For what?

File: helpers/uuid.go (hunk 2 above)
Comment (jenchik) [reply helpers/uuid.go]: you know it. eliminate uncertainties.

**Comment by im-kulikov** [comment]
```go
package helpers

import (
	"fmt"

	"github.com/satori/go.uuid"
)

// UUID representation compliant with specification
// described in RFC 4122.
type UUID = uuid.UUID

// NewUUID returns UUIDv4
func NewUUID() UUID {
	return uuid.NewV4()
}

// ValidateUUID from string
func ValidateUUID(id string) error {
	var (
		uid    UUID
		err    error
		length = len(id)
	)
	if length < uuid.Size*2 {
		return fmt.Errorf("uuid: invalid length(`%s`): %v", id, length)
	}
	if uid, err = uuid.FromString(id); err != nil {
		return fmt.Errorf("uuid: validation error(`%s`): %v", id, err)
	}
	if uid.Version() != 4 {
		return fmt.Errorf("uuid: invalid version number (`%s`): %d", id, uid.Version())
	}
	return nil
}

// ValidateUUIDs v4 from slice of string
func ValidateUUIDs(uuids []string) (err error) {
	if len(uuids) == 0 {
		return
	}
	for _, id := range uuids {
		if err = ValidateUUID(id); err != nil {
			return
		}
	}
	return
}
```

Why not like this? and you not need any tests around the wrappers

**Comment by im-kulikov** [comment]
```go
package helpers_test

import (
	"math/rand"
	"testing"

	. "github.com/cryptopay-dev/yaga/helpers"
	"github.com/satori/go.uuid"
	"github.com/stretchr/testify/assert"
)

var uuidCase = []struct {
	uuid string
	result bool
}
{
	{
		uuid:   "",
		result: false,
	},
	{
		uuid:   "empty",
		result: false,
	},
	{
		uuid:   uuid.NewV1().String(),
		result: false,
	},
	{
		uuid:   "b17c7f4b981e43658679d16d5837a7eb",
		result: true,
	},
	{
		uuid:   "b17c7f4b-981e-4365-8679-d16d5837a7eb",
		result: true,
	},
	{
		uuid:   "{b17c7f4b-981e-4365-8679-d16d5837a7eb}",
		result: true,
	},
	{
		// not 4 version
		uuid:   "b17c7f4b-981e-7365-8679-d16d5837a7eb",
		result: false,
	},
	{
		uuid:   "b17c7f4b-981e-4365-8x79-d16d5837a7eb",
		result: false,
	},
	{
		uuid:   "k17c7f4b-981e-4365-8679-d16d5837a7eb",
		result: false,
	},
}

var uuidsCase = []struct {
	uuids  []string
	result bool
}{
	{
                // Valid UUIDv1
		uuids: []string{
			"b94caa7a-0fdb-11e8-b642-0ed5f89f718b",
			"b94cadea-0fdb-11e8-b642-0ed5f89f718b",
			"b94cafd4-0fdb-11e8-b642-0ed5f89f718b",
			"b94cb25e-0fdb-11e8-b642-0ed5f89f718b",
			"b94cb510-0fdb-11e8-b642-0ed5f89f718b",
			"b94cb7b8-0fdb-11e8-b642-0ed5f89f718b",
			"b94cbd58-0fdb-11e8-b642-0ed5f89f718b",
			"b94cc00a-0fdb-11e8-b642-0ed5f89f718b",
			"b94cc2b2-0fdb-11e8-b642-0ed5f89f718b",
			"b94cc5e6-0fdb-11e8-b642-0ed5f89f718b",
		},
		result: false,
	},
	{
                // Valid UUIDv1 and one v4
		uuids: []string{
			"b94caa7a-0fdb-11e8-b642-0ed5f89f718b",
			"b94cadea-0fdb-11e8-b642-0ed5f89f718b",
			"b94cafd4-0fdb-11e8-b642-0ed5f89f718b",
			"b94cb25e-0fdb-11e8-b642-0ed5f89f718b",
			"b94cb510-0fdb-11e8-b642-0ed5f89f718b",
			"b94cb7b8-0fdb-11e8-b642-0ed5f89f718b",
			"b94cbd58-0fdb-11e8-b642-0ed5f89f718b",
			"b94cc00a-0fdb-11e8-b642-0ed5f89f718b",
			"b94cc2b2-0fdb-11e8-b642-0ed5f89f718b",
			"5aaad328-5d5b-4ff8-bb65-3bc3de8e2f2f", // v4
		},
		result: false,
	},
	{
                // Empty
		uuid:   []string{},
		result: true,
	},
	{
                // Valid UUIDv4 and one bad
		uuid:   []string{
			"",
			"9e5cdce4-b75a-4c61-910d-136492121971",
			"925294c6-bb3b-4d80-9f63-a634ae67264c",
			"07f06cf5-8bc6-4440-aa90-e2e166dbac45",
			"5b12eef1-a38c-4763-a872-d5b8be0f5849",
			"35f5e09e-ef7a-4075-91f4-1cecc72313c1",
			"df6179b1-488a-4e8d-ab0b-b18815ae5305",
			"e7d96ec4-07ba-4ce0-a85b-3f52d4907bf6",
			"f0654774-9185-48f9-bfd9-9114df958cc6",
			"5aaad328-5d5b-4ff8-bb65-3bc3de8e2f2f",
		},
		result: true,
	},
	{
                // Valid UUIDv4 and one bad
		uuid:   []string{
			"9e5cdce4-b75a-4c61-910d-136492121971",
			"925294c6-bb3b-4d80-9f63-a634ae67264c",
			"07f06cf5-8bc6-4440-aa90-e2e166dbac45",
			"5b12eef1-a38c-4763-a872-d5b8be0f5849",
			"35f5e09e-ef7a-4075-91f4-1cecc72313c1",
			"df6179b1-488a-4e8d-ab0b-b18815ae5305",
			"e7d96ec4-07ba-4ce0-a85b-3f52d4907bf6",
			"f0654774-9185-48f9-bfd9-9114df958cc6",
			"5aaad328-5d5b-4ff8-bb65-3bc3de8e2f2f",
			"not-an-uuid-v4",
		},
		result: true,
	},
	{
                // Valid UUIDv4
		uuid:   []string{
			"16822dd1-0018-4dd8-8d1d-4f7b071ddfd4",
			"9e5cdce4-b75a-4c61-910d-136492121971",
			"925294c6-bb3b-4d80-9f63-a634ae67264c",
			"07f06cf5-8bc6-4440-aa90-e2e166dbac45",
			"5b12eef1-a38c-4763-a872-d5b8be0f5849",
			"35f5e09e-ef7a-4075-91f4-1cecc72313c1",
			"df6179b1-488a-4e8d-ab0b-b18815ae5305",
			"e7d96ec4-07ba-4ce0-a85b-3f52d4907bf6",
			"f0654774-9185-48f9-bfd9-9114df958cc6",
			"5aaad328-5d5b-4ff8-bb65-3bc3de8e2f2f",
		},
		result: true,
	},
}

func TestNewUUIDv4(t *testing.T) {
	for i := 0; i < 10; i++ {
		assert.NoError(t, ValidateUUIDv4(NewUUIDv4()))
	}

	assert.NoError(t, ValidateUUIDv4("{"+NewUUIDv4()+"}"))
	assert.Error(t, ValidateUUIDv4("["+NewUUIDv4()+"]"))
}

func TestValidateUUIDv4(t *testing.T) {
	for _, suite := range uuidCase {
		if suite.result {
			assert.NoError(t, ValidateUUIDv4(suite.uuid))
		} else {
			assert.Error(t, ValidateUUIDv4(suite.uuid))

		}
	}
}

func TestValidateUUIDs(t *testing.T) {
	for _, suite := range uuidsCase {
		if suite.result {
			assert.NoError(t, ValidateUUIDv4(suite.uuid))
		} else {
			assert.Error(t, ValidateUUIDv4(suite.uuid))

		}
	}
}
```

**Review by kpeo — DISMISSED** [review]

File: helpers/uuid.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		return err
+	} else if uid.Version() != 4 {
+		return fmt.Errorf("uuid: invalid version number, must be '4', actual '%d'", uid.Version())
+	}
+
+	return nil
+}
+
+// ValidateUUIDs validation UUIDs for conformity with version 4
+func ValidateUUIDs(uuids []string) (err error) {
```
Comment (im-kulikov) [inline helpers/uuid.go:31]: here...

File: helpers/uuid.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// NewUUIDv4 returns random generated canonical string representation of UUID:
+// xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.
+func NewUUIDv4() string {
+	return uuid.NewV4().String()
+}
+
+// ValidateUUIDv4 verification UUID as string for conformity with version 4
+func ValidateUUIDv4(src string) error {
```
Comment (im-kulikov) [inline helpers/uuid.go]: here

File: helpers/uuid.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"fmt"
+
+	"github.com/satori/go.uuid"
+)
+
+// NewUUIDv4 returns random generated canonical string representation of UUID:
+// xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.
+func NewUUIDv4() string {
```
Comment (im-kulikov) [inline helpers/uuid.go]: here

**Review by im-kulikov — APPROVED** [review]

**Review by pcherednichenko — APPROVED** [review]

**Pull request merged**